gaps/gap-24_agreement_permissions_management/gap-24_agreement_permissions_management.md

```diff
@@ -0,0 +1,269 @@
+
+Therefore a rights Grantee, who is willing to Accept an incoming Debit Note or Invoice must pass `allocationId` of their own Allocation.
+
+### Transitive permissions
+
+**Note** granting access to `req:setAgreementPermissions` effectively allows grantee to delegates access rights further. This implies transitivity of permissions.
+
+### Implementation considerations
+
+To implement the Agreement Permissions Management it is necessary to introduce permission control in yagna REST API. However there is a number of possible ways to manage and 'guard' the Agreement's ACL.
+
+1. ACL managed and exerted on Provider node
+
+In this model, the ACL is persisted and managed on the Provider node indicated by the Agreement. Therefore all actions resulting from REST API calls must be routed to Provider node, and the permissionsneed to be validated there. Any attempt to violate the permissions as indicated by the ACL must result with "permissions violation" error returned to the calling node (and respective REST API caller).
+
+The benefit of this model is the fact that ACL is managed and verified in one place, which limits the possibility of race conditions, therefore increasing the robustness of implementation.
+
+The drawback is - all actions must be routed to the Provider for validation, and that increases delay before a REST API call result is returned to the caller. 
+
+2. ACL distributed and exerted on Requestor and Grantee nodes
+
+In this model, the ACL is managed by the Provider node indicated by the Agreement, however each change in the ACL is distributed to all Grantee nodes. This allows for the permission verification to happen on Grantee nodes.
+
+The benefit is - immediate verification of permissions, as each Grantee node can verify the permissions against their copy of the ACL.
+
+The drawback is the risk of local ACL copies going out of sync versus theProvider's master-copy. This introduces a risk of race conditions, and generally reduces the security level of the mechanism.
+
+**Recommendation:** Implementation 1. 
+
+## Rationale
+
+### Permissions API?
+Permission management enhancements are to be applied on Market API only, as only Agreement ACLs are to be managed. An alternative is a broad, generic Permissions API, but it seems overkill.
+
+### ACL structure
+A simple ACL structure was considered:
+
+```
+agreementACL = {
+    version: 123,
+    aclEntries: [
+        {
+            "grantee": "0x7735df0c0bbb3ca65c55d61eadded653573e0152",
+            "permissions" : [
+                "req:createActivity",
+                "req:exec",
+                "req:destroyActivity"
+            ]
+        },
+        ...
+        {
+            "grantee": "0x0053b2c00006d1d8b75bb7dfa068990a19e4c32e",
+            "permissions" : [
+                "req:terminateAgreement",
+                "req:acceptInvoice",
```

Comment:
I realized that accepting Invoice (and in result making payment) is something that won't be processed by Provider (he only gets event). This means that permissions are no guarantee that the payment won't be done, so our blockchain contract must be consensus aware

Reply:
Hmmm, actually I don't think that would be the case. Note that we agreed any action which may change state of any entity (like Invoice) must be verified (permission-wise) on Provider side to avoid race conditions. Only after the Provider confirms `acceptInvoice` succeeds - the payment stack on Requestor side may queue the payment.

Reply:
This means that Provider is only expected to be a "synchronization point" for Invoice state change, and nothing else (ie. the Payment still needs to be orchestrated by Requestor, after Invoice state change is confirmed by Provider).